src/visitable.cpp

```diff
@@ -212,6 +213,249 @@ VisitResponse visitable<vehicle_selector>::visit_items_with_parent(
+            res.splice( res.end(), ch->worn, iter++ );
+            if( --count == 0 ) {
+                return res;
+            }
+        } else {
+            remove_internal( filter, *iter, count, std::back_inserter( res ) );
+            if( count == 0 ) {
+                return res;
+            }
+            ++iter;
+        }
+    }
+
+    // finally try the currently wielded item (if any)
+    if( filter( ch->weapon ) ) {
+        res.push_back( ch->remove_weapon() );
+        count--;
+    } else {
+        remove_internal( filter, ch->weapon, count, std::back_inserter( res ) );
+    }
+
+    return res;
+}
+
+template <>
+std::list<item> visitable<map_cursor>::remove_items_with( const
+        std::function<bool( const item &e )> &filter, int count )
+{
+    auto cur = static_cast<map_cursor *>( this );
+    std::list<item> res;
+
+    if( count == 0 ) {
+        return res; // nothing to do
+    }
+
+    if( !g->m.inbounds( *cur ) ) {
+        debugmsg( "cannot remove items from map: cursor out-of-bounds" );
+        return res;
+    }
+
+    // fetch the appropriate item stack
+    int x, y;
+    submap *sub = g->m.get_submap_at( *cur, x, y );
+
+    for( auto iter = sub->itm[ x ][ y ].begin(); iter != sub->itm[ x ][ y ].end(); ) {
+        if( filter( *iter ) ) {
+            // check for presence in the active items cache
+            if( sub->active_items.has( iter, point( x, y ) ) ) {
+                sub->active_items.remove( iter, point( x, y ) );
+            }
+
+            // if necessary remove item from the luminosity map
+            sub->update_lum_rem( *iter, x, y );
+
+            // finally remove the item
+            res.splice( res.end(), sub->itm[ x ][ y ], iter++ );
+
+            if( --count == 0 ) {
+                return res;
+            }
+        } else {
+            remove_internal( filter, *iter, count, std::back_inserter( res ) );
+            if( count == 0 ) {
+                return res;
+            }
+            ++iter;
+        }
+    }
+    return res;
+}
+
+template <>
+std::list<item> visitable<map_selector>::remove_items_with( const
+        std::function<bool( const item &e )> &filter, int count )
+{
+    std::list<item> res;
+
+    for( auto &cursor : static_cast<map_selector &>( *this ) ) {
+        std::list<item> out = cursor.remove_items_with( filter, count );
+        count -= out.size();
+        res.splice( res.end(), out );
+    }
+
+    return res;
+}
+
+template <>
+std::list<item> visitable<vehicle_cursor>::remove_items_with( const
+        std::function<bool( const item &e )> &filter, int count )
+{
+    auto cur = static_cast<vehicle_cursor *>( this );
+    std::list<item> res;
+
+    if( count == 0 ) {
+        return res; // nothing to do
+    }
+
+    vehicle_part& part = cur->veh.parts[ cur->part ];
+    for( auto iter = part.items.begin(); iter != part.items.end(); ) {
+        if( filter( *iter ) ) {
+            // check for presence in the active items cache
```

Comment:
Unless you object to the extra code review I might actually push the trivial implementation of `remove_item` as part of this PR?